book/src/internals/name-resolving.md

```diff
@@ -0,0 +1,77 @@
+# Name resolving
+
+Because PRQL primarily handles relational data, it has specialized scoping rules for referencing columns.
+
+## Scopes
+
+In PRQL's compiler, a scope is the collection of all names one can reference from a specific point in the program.
+
+In PRQL, names in the scope are composed from namespace and variable name which are separated by a dot, similar to SQL.
+Namespaces can contain many dots, but variable names cannot.
+
+```admonish example
+Name `my_table.some_column` is a variable `some_column` from namespace `my_table`.
+
+Name `foo.bar.baz` is a variable `baz` from namespace `foo.bar`.
+```
+
+When processing a query, a scope is maintained and updated for each point in the query.
+
+It start with only namespace `std`, which is the standard library.
+It contains common functions like `sum` or `count`,
+along with all transform functions such as `derive` and `group`.
+
+In pipelines (or rather in transform functions),
+scope is also injected with namespaces of tables which may have been referenced with `from` or `join` transforms.
+These namespaces contain simply all the columns of the table and possibly a wildcard variable,
+which matches any variable (see the algorithm below).
+Within transforms, there is also a special namespace that does not have a name.
+It is called a _"frame"_ and it contains columns of the current table the transform is operating on.
+
+## Resolving
+
+For each ident we want to resolve, we search the scope's items in order. One of three things can happen:
```

Comment:
What should we do with expressions that only include a table name? This might only be relevant for S-strings, but I mean something like this: `s"{tablethatprqlknows}.somefield"`

If the `tablethatprqlknows` table is in scope, then this should just resolve to that

There's another example showing when this can be important here: https://github.com/prql/prql/issues/820#issuecomment-1325652225 (though, I'm not sure if I can bring up an example use case once #820 is fixed)